File: SeverTest/Modules/StrategyEngine/Application/TalibCalcRules.cs
using ccxt;

namespace ServerTest.Modules.StrategyEngine.Application
{
    /// <summary>
    /// TA 指标统一计算规则。
    /// 该规则需要与前端 talibCalcRules.ts 保持一致。
    /// </summary>
    internal static class TalibCalcRules
    {
        /// <summary>
        /// 指标最小历史根数下限，避免空窗口导致计算失败。
        /// </summary>
        private const int MinRequiredBars = 5;

        /// <summary>
        /// EMA 族指标的最小预热根数（用于降低短窗口初值偏移）。
        /// </summary>
        private const int EmaFamilyMinWarmupBars = 200;

        public static int RoundToIntOption(double value)
        {
            if (double.IsNaN(value) || double.IsInfinity(value))
            {
                return 0;
            }

            return (int)Math.Round(value, MidpointRounding.AwayFromZero);
        }

        public static string NormalizeInputSource(string? source)
        {
            var normalized = (source ?? string.Empty)
                .Trim()
                .ToUpperInvariant();

            return normalized switch
            {
                "OPEN" => "OPEN",
                "HIGH" => "HIGH",
                "LOW" => "LOW",
                "CLOSE" => "CLOSE",
                "VOLUME" => "VOLUME",
                "HL2" => "HL2",
                "HLC3" => "HLC3",
                "OHLC4" => "OHLC4",
                "OC2" => "OC2",
                "HLCC4" => "HLCC4",
                _ => "CLOSE"
            };
        }

        public static double ResolveSourceValue(OHLCV candle, string? source)
        {
            var key = NormalizeInputSource(source);
            var open = candle.open ?? double.NaN;
            var high = candle.high ?? double.NaN;
            var low = candle.low ?? double.NaN;
            var close = candle.close ?? double.NaN;
            var volume = candle.volume ?? double.NaN;

            return key switch
            {
                "OPEN" => open,
                "HIGH" => high,
                "LOW" => low,
                "CLOSE" => close,
                "VOLUME" => volume,
                "HL2" => (high + low) / 2.0,
                "HLC3" => (high + low + close) / 3.0,
                "OHLC4" => (open + high + low + close) / 4.0,
                "OC2" => (open + close) / 2.0,
                "HLCC4" => (high + low + close + close) / 4.0,
                _ => close
            };
        }

        /// <summary>
        /// 统一计算指标所需历史根数。
        /// 规则：
        /// 1) 基础规则：lookback + maxOffset + 2；
        /// 2) EMA 族规则：额外要求至少 200 根预热，降低初值偏移。
        /// </summary>
        public static int ResolveRequiredBars(string? indicator, int lookback, int maxOffset)
        {
            var safeLookback = Math.Max(0, lookback);
            var safeMaxOffset = Math.Max(0, maxOffset);
            var baseRequired = Math.Max(MinRequiredBars, safeLookback + safeMaxOffset + 2);
            if (!IsEmaFamily(indicator))
            {
                return baseRequired;
            }

            var emaRequired = safeMaxOffset + EmaFamilyMinWarmupBars;
            return Math.Max(baseRequired, emaRequired);
        }

        private static bool IsEmaFamily(string? indicator)
        {
            var code = (indicator ?? string.Empty).Trim().ToUpperInvariant();
            return code switch
            {
                "EMA" => true,
                "DEMA" => true,
                "TEMA" => true,
                "TRIX" => true,
                "T3" => true,
                "KAMA" => true,
                "MAMA" => true,
                _ => false
            };
        }
    }
}

File: Client/src/lib/talibCalcRules.ts
import type { KLineData } from "klinecharts";

export type TalibInputSource =
  | "OPEN"
  | "HIGH"
  | "LOW"
  | "CLOSE"
  | "VOLUME"
  | "HL2"
  | "HLC3"
  | "OHLC4"
  | "OC2"
  | "HLCC4";

/**
 * 统一规范输入源名称，保证前后端使用一致的映射规则。
 */
export function normalizeTalibInputSource(value: string | null | undefined): TalibInputSource {
  const normalized = (value ?? "").trim().toUpperCase().replace(/[^A-Z0-9]/g, "");
  switch (normalized) {
    case "OPEN":
      return "OPEN";
    case "HIGH":
      return "HIGH";
    case "LOW":
      return "LOW";
    case "CLOSE":
      return "CLOSE";
    case "VOLUME":
      return "VOLUME";
    case "HL2":
      return "HL2";
    case "HLC3":
      return "HLC3";
    case "OHLC4":
      return "OHLC4";
    case "OC2":
      return "OC2";
    case "HLCC4":
      return "HLCC4";
    default:
      return "CLOSE";
  }
}

/**
 * 统一数值解析：无法解析为有限数字时保留 NaN，避免被静默改写为 0。
 */
export function toFiniteNumberOrNaN(value: unknown): number {
  if (typeof value === "number" && Number.isFinite(value)) {
    return value;
  }
  if (typeof value === "string") {
    const parsed = Number(value);
    return Number.isFinite(parsed) ? parsed : Number.NaN;
  }
  return Number.NaN;
}

/**
 * 对齐 C# MidpointRounding.AwayFromZero 的整数取整行为。
 */
export function roundAwayFromZero(value: number): number {
  if (!Number.isFinite(value) || value === 0) {
    return 0;
  }
  return Math.sign(value) * Math.round(Math.abs(value));
}

/**
 * 统一行情输入字段解析，所有派生源都基于同一套数学定义。
 */
export function resolveTalibSeriesValue(item: KLineData, source: TalibInputSource): number {
  const open = toFiniteNumberOrNaN(item.open);
  const high = toFiniteNumberOrNaN(item.high);
  const low = toFiniteNumberOrNaN(item.low);
  const close = toFiniteNumberOrNaN(item.close);
  const volume = toFiniteNumberOrNaN(item.volume);

  switch (source) {
    case "OPEN":
      return open;
    case "HIGH":
      return high;
    case "LOW":
      return low;
    case "CLOSE":
      return close;
    case "VOLUME":
      return volume;
    case "HL2":
      return (high + low) / 2;
    case "HLC3":
      return (high + low + close) / 3;
    case "OHLC4":
      return (open + high + low + close) / 4;
    case "OC2":
      return (open + close) / 2;
    case "HLCC4":
      return (high + low + close + close) / 4;
    default:
      return close;
  }
}
